### PR TITLE
Use Rust Loader for running examples

### LINKED ISSUE #725: Use Rust `oak_loader` for running examples
The examples may still be in C++, but we should run them against the Rust `oak_loader`, once it exists.

Perhaps we should add a `rust` value for the server type flag:

https://github.com/project-oak/oak/blob/d66958bba544ea97761749735f239020e67b8437/scripts/run_example#L19-L23

Blocked by #723

### LINKED ISSUE #874: Provide gRPC read handle to the Oak entrypoint
Function that runs an event loop for a **Node** is called in the `entrypoint` (aka `oak_main`) created by this macro:
https://github.com/project-oak/oak/blob/6712ee6ba89aa0cd93361b83aa0e5597540ba713/sdk/rust/oak/src/lib.rs#L567-L576
This entrypoint requires an `in_handle` to receive messages, and this handle is provided by the Runtime upon creating this **Node**.
In the case of gRPC server _pseudo-Node_ this `in_handle` is created during the _pseudo-Node_ creation. But this _pseudo-Node_ is created by the original **Node** with an entrypoint, that at this point has already been created and provided with an `in_handle`.
And this results in a loop.

We need to change a way of how the initial **Node** is created. 
We have several possible options:
- **Node** creates a gRPC server _pseudo-Node_, receives invocations from it, and dispatches them to the `impl` of the gRPC service, which is all in the same node.
- **Node** creates a gRPC server _pseudo-Node_, and the gRPC service node (which will still be declared using the current entrypoint macro), then forwards invocations to that node.
- Runtime creates two Nodes: a gRPC server _pseudo-Node_ and the main **Node** (which uses a `handle` returned by the former). This approach is similar to the current approach used in the C++ Runtime.

Originated from a Slack thread: https://project-oak.slack.com/archives/CHE9E13C3/p1587068768173500

cc @tiziano88 @daviddrysdale

### LINKED ISSUE #901: Running `hello_world` example doesn't include the translator
The Wasm module for the translator is built by `scripts/build_example -e hello_world` (as a special case), and two versions of the application config are built (`config.bin` and `config_translator.bin`), but `scripts/run_example -e hello_world` uses the version without the translator module included.

### PR DESCRIPTION
**Pull request by ipetr0v**
This change:
- Updates examples to use Rust Loader
- Updates macro and scripts for running Rust Loader
- Makes Oak CI to use Rust Loader for running examples
- Builds Rust Oak Loader with Cargo

Fixes #725
Fixes #874
Fixes #901
Ref #945

# Checklist
- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [x] I have updated [documentation](docs/) accordingly.
  - [x] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.

### CODE REVIEW

File: examples/aggregator/backend/src/main.rs [hunk 1]
```diff
@@ -28,7 +28,23 @@ use proto::{
...
+use tonic::{
+    transport::{Identity, Server, ServerTlsConfig},
+    Request, Response, Status,
+};
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Aggregator Backend")]
+pub struct Opt {
+    #[structopt(long, help = "Private RSA key file used by gRPC server.")]
+    tls_private_key: String,
```
Comment (tiziano88) [inline examples/aggregator/backend/src/main.rs:41]: Could you use the same flag names as for the `oak_loader`?

File: examples/aggregator/backend/src/main.rs [hunk 2]
```diff
@@ -48,12 +64,20 @@ impl Aggregator for AggregatorBackend {
 #[tokio::main]
 async fn main() -> Result<(), Box<dyn std::error::Error>> {
     env_logger::init();
-    let address = "127.0.0.1:8888".parse()?;
+    let opt = Opt::from_args();
+
+    let private_key = tokio::fs::read(&opt.tls_private_key).await?;
+    let certificate = tokio::fs::read(&opt.tls_certificate).await?;
```
Comment (tiziano88) [inline examples/aggregator/backend/src/main.rs:70]: Wow, so much parallelism :) (well, actually they are serialized anyways)

File: examples/aggregator/scripts/run_server [hunk 3]
```diff
@@ -5,11 +5,12 @@ readonly SCRIPTS_DIR="$(dirname "$(readlink -f "$0")")"
 # shellcheck source=scripts/common
 source "${GLOBAL_SCRIPTS_DIR}/common"
 
-"${GLOBAL_SCRIPTS_DIR}/build_server" -s base
+"${GLOBAL_SCRIPTS_DIR}/build_server" -s rust
+"${GLOBAL_SCRIPTS_DIR}/build_example" -e aggregator
 
 readonly APPLICATION="${PWD}/bazel-client-bin/examples/aggregator/config/config.bin"
-exec ./bazel-bin/oak/server/loader/oak_runner \
+exec ./bazel-clang-bin/oak/server/rust/oak_loader/oak_loader \
```
Comment (tiziano88) [inline examples/aggregator/scripts/run_server:12]: Wait, why are we not using the cargo version yet?

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs [hunk 4]
```diff
@@ -34,75 +34,74 @@ use tonic::transport::{Certificate, Channel, ClientTlsConfig, Uri};
...
+        handle: Handle,
+    ) -> Result<(), OakStatus> {
         // Create a [`Receiver`] used for reading gRPC invocations.
         let receiver = Receiver::<Invocation>::new(handle);
         loop {
+            debug!("Waiting for gRPC invocations");
             // Read a gRPC invocation from the [`Receiver`].
-            let invocation = receiver.receive(&runtime)?;
+            let invocation = receiver.receive(&runtime).map_err(|error| {
+                error!("Couldn't reveice the invocation: {:?}", error);
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/grpc/client.rs:69]:
```suggestion
                error!("Couldn't receive the invocation: {:?}", error);
```

File: sdk/rust/oak/src/lib.rs [hunk 5]
```diff
@@ -571,28 +576,31 @@ pub fn run_event_loop<T: crate::io::Decodable, N: Node<T>>(mut node: N, in_handl
...
-    ($name:ident => $node:expr) => {
+    ($name:ident => $main_function:expr) => {
         // Do not mangle these functions when running unit tests, because the Rust unit test
         // framework will add a `pub extern "C" fn main()` containing the test runner. This can
         // cause clashes when $name = main. We don't fully omit it in tests so that compile errors
         // in the Node creation expression are still caught, and unit tests can still refer to the
         // symbol if they really want to.
         #[cfg_attr(not(test), no_mangle)]
-        pub extern "C" fn $name(in_handle: u64) {
+        pub extern "C" fn $name(_: u64) {
```
Comment (tiziano88) [inline sdk/rust/oak/src/lib.rs:595]: Leave in the name with underscor. Also perhaps explain why this gets ignored.

File: examples/abitest/client/abitest.cc [hunk 6]
```diff
@@ -158,9 +160,30 @@ void run_grpc_test_server(int grpc_test_port, grpc::Server** grpc_test_server) {
...
+    OAK_LOG(FATAL) << "No certificate chain file specified.";
+  }
+  std::string private_key = oak::utils::read_file(private_key_path);
+  std::string cert_chain = oak::utils::read_file(cert_chain_path);
+
+  grpc::SslServerCredentialsOptions::PemKeyCertPair key_cert_pair = {private_key, cert_chain};
+  grpc::SslServerCredentialsOptions options;
+  options.pem_root_certs = ca_cert;
+  options.pem_key_cert_pairs.push_back(key_cert_pair);
+  return grpc::SslServerCredentials(options);
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply examples/abitest/client/abitest.cc:179]: Presumably this is more because the Rust gRPC *client* pseudo-Node now always uses TLS, so the test server that it connects to (which runs in parallel inside the `abitest` client program) now needs to act as TLS server side?

File: examples/running_average/client/running_average.cc [hunk 7]
```diff
@@ -84,9 +86,11 @@ int main(int argc, char** argv) {
   // Retrieve average.
   int average_0 = retrieve_average(stub_0.get());
   LOG(INFO) << "client 0 average: " << average_0;
+  assert(average_0 == 100);
```
Comment (daviddrysdale) [inline examples/running_average/client/running_average.cc:89]: This looks like an unrelated change?

File: examples/abitest/config/config.textproto [hunk 8]
```diff
@@ -26,17 +26,27 @@ node_configs {
     address: "test.invalid:9999"
   }
 }
+node_configs {
```
Comment (daviddrysdale) [inline examples/abitest/config/config.textproto:29]: Is it possible to update [`oak::DefaultConfig()`](https://github.com/project-oak/oak/blob/master/oak/common/app_config.cc#L44-L56) to include something like this stanza, if it's now effectively required?  Or is that not possible because of the need to know a TLS auth data in advance?

File: sdk/rust/oak/src/lib.rs [hunk 9]
```diff
@@ -517,8 +518,8 @@ pub fn run_event_loop<T: crate::io::Decodable, N: Node<T>>(mut node: N, in_handl
 
 /// Register a new Node entrypoint.
 ///
-/// This registers the entrypoint name and the expression used to construct the
-/// Node instance. The returned object should implement the [`Node`](trait.Node.html) trait.
+/// This registers the entrypoint name and the expression used to construct the Node instance.
+/// The returned object should be a function that runs [`run_event_loop`].
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/lib.rs:522]: These 2 sentences need a reword for the new way of doing things (there's nothing returned, and in particular not a `Node` instance returned).

File: sdk/rust/oak/src/lib.rs [hunk 10]
```diff
@@ -538,7 +539,11 @@ pub fn run_event_loop<T: crate::io::Decodable, N: Node<T>>(mut node: N, in_handl
 ///     # }
 /// }
 ///
-/// oak::entrypoint!(dummy => DummyNode::default());
+/// oak::entrypoint!(dummy => {
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/lib.rs:542]: This style of macro looks weird when the body involved is now just the body of the main function run by the Node.  I think this should revert to a more "normal" macro style:
```Rust
#[oak_grpc_main]
fn dummy_front() {  // no Handle available for initial node
    let dispatcher = FrontendNode::default();
    let grpc_channel = oak::grpc::server::init_default();
    oak::run_event_loop(dispatcher, grpc_channel);
}
#[oak_main]
fn dummy_inner(handle: oak::ReadHandle) {
    let dispatcher = BackendNode::default();
    oak::run_event_loop(dispatcher, handle);
}
```
(For a follow-up).

File: examples/translator/module/rust/src/lib.rs [hunk 11]
```diff
@@ -25,7 +25,9 @@ use translator_common::proto::{
 
 oak::entrypoint!(oak_main => {
     oak::logger::init_default();
-    TranslatorDispatcher::new(Node)
+    let dispatcher = TranslatorDispatcher::new(Node);
+    let grpc_channel = oak::grpc::server::init_default();
+    oak::run_event_loop(dispatcher, grpc_channel);
```
Comment (daviddrysdale) [inline examples/translator/module/rust/src/lib.rs:30]: This will need to be split up somehow, so it's still possible to use the `translator` Node as a library Node that's part of another Application &ndash; in the latter context, the Node instance will be handed an initial handle and *shouldn't* create a new gRPC server Node.

(If #901 weren't open, the `hello_world` example would illustrate this use case.)

File: scripts/run_examples [hunk 12]
```diff
@@ -4,15 +4,16 @@ readonly SCRIPTS_DIR="$(dirname "$0")"
 # shellcheck source=scripts/common
 source "${SCRIPTS_DIR}/common"
 
-server="base"
+server="rust"
 languages="rust cpp"
 while getopts "s:l:h" opt; do
   case "${opt}" in
     h)
       echo -e "Usage: ${0} [-s base|logless|asan|tsan] [-l rust|cpp]
```
Comment (daviddrysdale) [inline scripts/run_examples:12]: nit: update usage message for `rust` option

File: scripts/run_server [hunk 13]
```diff
@@ -5,7 +5,7 @@ readonly SCRIPTS_DIR="$(dirname "$0")"
 source "${SCRIPTS_DIR}/common"
 
 language="rust"
-server="base"
+server="rust"
 while getopts "s:a:l:e:vh" opt; do
   case "${opt}" in
     h)
```
Comment (daviddrysdale) [inline scripts/run_server:11]: nit, pre-existing: update usage statement on line below to include `rust`

File: scripts/run_example [hunk 14]
```diff
@@ -18,9 +18,9 @@ Build and run the given example Oak Application and client.
 Options:
```
Comment (daviddrysdale) [inline scripts/run_example:18]: nit, pre-existing: update usage message 4 lines up to include `rust`

File: oak/server/rust/oak_runtime/src/node/storage.rs [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+impl super::Node for StorageNode {
+    fn run(
+        self: Box<Self>,
+        _runtime: RuntimeProxy,
+        _handle: oak_abi::Handle,
+        _notify_receiver: oneshot::Receiver<()>,
+    ) {
+        info!("{}: Starting storage pseudo-Node stub", self.node_name);
```
Comment (daviddrysdale) [inline oak/server/rust/oak_runtime/src/node/storage.rs:42]: nit: Be nice to add a `warn!("No storage implemented!")` just so logs are clear.

File: oak/server/rust/oak_runtime/src/node/roughtime.rs [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+impl super::Node for RoughtimeClientNode {
+    fn run(
+        self: Box<Self>,
+        _runtime: RuntimeProxy,
+        _handle: oak_abi::Handle,
+        _notify_receiver: oneshot::Receiver<()>,
+    ) {
+        info!("{}: Starting Roughtime pseudo-Node stub", self.node_name);
```
Comment (daviddrysdale) [inline oak/server/rust/oak_runtime/src/node/roughtime.rs:42]: nit: Be nice to add a `warn!("No roughtime support implemented!");` just so logs are clear.

File: oak/server/rust/oak_runtime/src/node/mod.rs [hunk 17]
```diff
@@ -206,6 +219,10 @@ impl Configuration {
             Configuration::GrpcServerNode { .. } => "GrpcServerNode".to_string(),
             Configuration::GrpcClientNode { .. } => "GrpcClientNode".to_string(),
             Configuration::WasmNode { .. } => format!("WasmNode-{}", entrypoint),
+            Configuration::StorageNode { .. } => format!("StorageNode-{}", entrypoint),
+            Configuration::RoughtimeClientNode { .. } => {
+                format!("RoughtimeClientNode-{}", entrypoint)
+            }
```
Comment (daviddrysdale) [inline oak/server/rust/oak_runtime/src/node/mod.rs:225]: Best to drop the -*entrypoint* suffix; as with the other pseudo-Node types, the entrypoint param is ignored for these.

File: examples/running_average/client/running_average.cc (hunk 7 above)
Comment (ipetr0v) [reply examples/running_average/client/running_average.cc:89]: Moved to https://github.com/project-oak/oak/pull/1006

File: examples/private_set_intersection/client/private_set_intersection.cc [hunk 18]
```diff
@@ -87,18 +89,22 @@ int main(int argc, char** argv) {
...
 
+  std::set<std::string> expected_set{"b", "c"};
+
   // Retrieve intersection.
   std::vector<std::string> intersection_0 = RetrieveIntersection(stub_0.get());
   LOG(INFO) << "client 0 intersection:";
   for (auto item : intersection_0) {
     LOG(INFO) << "- " << item;
   }
+  assert(std::set<std::string>(intersection_0.begin(), intersection_0.end()) == expected_set);
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply examples/private_set_intersection/client/private_set_intersection.cc:100]: Moved to https://github.com/project-oak/oak/pull/1006

File: examples/translator/module/rust/src/lib.rs (hunk 11 above)
Comment (ipetr0v) [reply examples/translator/module/rust/src/lib.rs:30]: Fixed

File: sdk/rust/oak/src/lib.rs (hunk 10 above)
Comment (ipetr0v) [reply sdk/rust/oak/src/lib.rs:541]: Fixed
Translator now looks like this:
```rust
oak::entrypoint!(oak_main => |_in_channel| {
    oak::logger::init_default();
    let dispatcher = TranslatorDispatcher::new(Node);
    let grpc_channel = oak::grpc::server::init_default();
    oak::run_event_loop(dispatcher, grpc_channel);
});

oak::entrypoint!(backend_oak_main => |in_channel| {
    oak::logger::init_default();
    let dispatcher = TranslatorDispatcher::new(Node);
    oak::run_event_loop(dispatcher, in_channel);
});
```

File: examples/abitest/config/config.textproto (hunk 8 above)
Comment (tiziano88) [reply examples/abitest/config/config.textproto:29]: This is going away soon anyways.

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs [hunk 19]
```diff
@@ -137,37 +136,79 @@ impl Node for GrpcClientNode {
...
+
     /// Checks whether the [`GrpcRequestHandler::dispatcher`] is able to accept new requests.
     async fn ready(&mut self) -> Result<(), OakStatus> {
-        self.dispatcher.ready().await.map_err(|error| {
-            error!("Service was not ready: {}", error);
-            OakStatus::ErrInternal
-        })?;
+        self.dispatcher
+            .as_mut()
+            .expect("It not connected to service")
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/grpc/client.rs:186]: What is "It" here?

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs [hunk 20]
```diff
@@ -137,37 +136,79 @@ impl Node for GrpcClientNode {
...
         let codec = VecCodec::default();
         let path = request.method_name.parse().map_err(|error| {
             error!("Invalid URI {}: {}", request.method_name, error);
             OakStatus::ErrInternal
         })?;
 
+        // Send gRPC request and wait for the response.
         self.dispatcher
+            .as_mut()
+            .expect("Is not connected")
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/grpc/client.rs:211]: What is "Is" here?

File: Cargo.lock [hunk 21]
```diff
@@ -86,8 +86,9 @@ dependencies = [
  "futures-util 0.3.5 (registry+https://github.com/rust-lang/crates.io-index)",
  "log 0.4.8 (registry+https://github.com/rust-lang/crates.io-index)",
  "prost 0.6.1 (registry+https://github.com/rust-lang/crates.io-index)",
+ "structopt 0.3.14 (registry+https://github.com/rust-lang/crates.io-index)",
  "tokio 0.2.20 (registry+https://github.com/rust-lang/crates.io-index)",
- "tonic 0.2.0 (registry+https://github.com/rust-lang/crates.io-index)",
+ "tonic 0.1.1 (registry+https://github.com/rust-lang/crates.io-index)",
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply Cargo.lock:91]: There were different versions of `tonic` in `Cargo.lock`

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs (hunk 19 above)
Comment (ipetr0v) [reply oak/server/rust/oak_runtime/src/node/grpc/client.rs:191]: Changed to `Handler is not connected to a gRPC service`

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs (hunk 20 above)
Comment (ipetr0v) [reply oak/server/rust/oak_runtime/src/node/grpc/client.rs:216]: Changed to `Handler is not connected to a gRPC service`
